### PR TITLE
core: fix some "call on nullable" warnings

### PR DESCRIPTION
**Pull request by Morriar**
Too tired too work but no enough to sleep...

According to Jenkins [here](http://gresil.org/jenkins/job/CI_github/4177/warnings2Result/package.795683049/), [here](http://gresil.org/jenkins/job/CI_github/4177/warnings2Result/package.-1960060028/) and [here](http://gresil.org/jenkins/job/CI_github/4177/warnings2Result/package.1056020115/), this PR should fix the remaining warning of `core/`.

The algorithm used for the human/pillow approach was:

~~~
1. can it be fixed shamelessly with a .as(not null) ?
    yes: add that sweet .as(not null) and hope for the best
2. is there more than one occurrence of the variable involved?
    yes: factorise with a local variable, add that sweet .as(not null) and hope for the best
3. is this code to complex for my human/pillow head?
    yes: shame = false, fallback to 1.
abort "Not yet occurred case..."
~~~

There is love for @R4PaSs (core::text), @xymus (core::re) and @privat (core::*).

TL;DR No more warnings in core.

### CODE REVIEW

File: lib/core/file.nit [hunk 1]
```diff
@@ -49,23 +49,23 @@ abstract class FileStream
...
 
 	# File descriptor of this file
-	fun fd: Int do return _file.fileno
+	fun fd: Int do return _file.as(not null).fileno
 
 	redef fun close
 	do
 		if _file == null then return
-		if _file.address_is_null then
+		if _file.as(not null).address_is_null then
```
Comment (privat) [inline lib/core/file.nit]: there is a `_file == null` just above, what about a local variable instead?

File: lib/core/text/abstract_text.nit [hunk 2]
```diff
@@ -2021,7 +2021,7 @@ end
 # see `alpha_comparator`
 private class AlphaComparator
 	super Comparator
-	redef fun compare(a, b) do return a.to_s <=> b.to_s
+	redef fun compare(a, b) do return a.as(not null).to_s <=> b.as(not null).to_s
```
Comment (privat) [inline lib/core/text/abstract_text.nit]: Maybe allow to compare arrays that contains null ?

~~~nit
if a == b then return 0
if a == null then return -1
if b == null then return 1
return a <=> b
~~~

File: lib/core/text/ropes.nit [hunk 3]
```diff
@@ -936,26 +936,26 @@ private class RopeSubstrings
...
-				it.rdone = true
+			else if not it.as(not null).rdone then
+				it.as(not null).rdone = true
 				rnod = rnod._right
 				it = new RopeCharIteratorPiece(rnod, false, false, it)
 			else
-				it = it.prev
-				rnod = it.node
+				it = it.as(not null).prev
+				rnod = it.as(not null).node
```
Comment (privat) [inline lib/core/text/ropes.nit]: too much assnotnull I think. Why `it` cannot be a non-nullable variable?

**Comment by privat** [comment]
+1. But I'm not sure of what you broke in the tests.

File: lib/core/text/abstract_text.nit (hunk 2 above)
Comment (Morriar) [inline lib/core/text/abstract_text.nit]: I see this as a change of specification, thus beyond the scope of this PR?

File: lib/core/text/ropes.nit (hunk 3 above)
Comment (Morriar) [inline lib/core/text/ropes.nit]: Because with the current code, `it` must stay nullable since `it = it.prev`. It required a lot of code modification, and it was 3 a.m., I was not willing to do it. 

Now, at 6 p.m., I look at the name of the module and HELL NO, I'm not touching that code :p

File: lib/core/text/abstract_text.nit (hunk 2 above)
Comment (privat) [inline lib/core/text/abstract_text.nit]: I dont know. this removes warnings an make the code more robust. No existing valid case are changed. You did basically the same thing with `Set::hash` that does not crash on null elements

**Comment by Morriar** [comment]
voodoo toaster: test this please

File: lib/core/collection/array.nit [hunk 4]
```diff
@@ -475,8 +475,8 @@ class Array[E]
 		var l = length
 		if l != o.length then return false
 		var i = 0
-		var it = _items
-		var oit = o._items
+		var it = _items.as(not null)
+		var oit = o._items.as(not null)
 		while i < l do
```
Comment (privat) [inline lib/core/collection/array.nit]: the case where the length is 0 may crash.
Add a `if l == 0 then return true` after the `if l != o.length then return false`

File: lib/core/collection/array.nit [hunk 5]
```diff
@@ -921,8 +921,8 @@ class ArrayCmp[E: nullable Comparable]
 
 	redef fun <=>(o)
 	do
-		var it = _items
-		var oit = o._items
+		var it = _items.as(not null)
+		var oit = o._items.as(not null)
```
Comment (privat) [inline lib/core/collection/array.nit]: same length == 0 case here.

**Comment by Morriar** [comment]
Reroll, fixed all @privat remarks, trying to pass ~~the gates of hell~~ Jenkins tests.

**Comment by privat** [comment]
image

**Comment by Morriar** [comment]
Reroll: fixed line numbers in tests.

Should be good after that.

**Comment by Morriar** [comment]
toaster: test this please

**Comment by Morriar** [comment]
RIP Jenkins... :(

**Comment by privat** [comment]
toaster: test this please

**Comment by privat** [comment]
you need the right sentence

File: lib/core/collection/array.nit [hunk 6]
```diff
@@ -921,10 +922,11 @@ class ArrayCmp[E: nullable Comparable]
...
 		var i = 0
 		var l = length
+		if l == 0 then return 0
```
Comment (privat) [inline lib/core/collection/array.nit]: too late. should be before calling `_items`

Comment (Morriar) [inline lib/core/collection/array.nit]: But tests are ok :p

**Comment by Morriar** [comment]
@privat: reroll
